File: output_20260319_145930/qwen3/c/Stack/filter.c
#include "stdlib.h"

/*@ predicate node(struct node *n; int value, struct node *next) =
    n != 0 &*&
    malloc_block_node(n) &*&
    n->value |-> value &*&
    n->next |-> next;
@*/

/*@ predicate stack(struct stack *s; list<int> values) =
    s != 0 &*&
    malloc_block_stack(s) &*&
    s->head |-> ?head &*&
    nodes(head, values);
@*/

/*@ predicate nodes(struct node *n; list<int> values) =
    switch (values) {
        case nil: return n == 0;
        case cons(h, t): return node(n, h, ?next) &*& nodes(next, t);
    };
@*/

struct node
{
    struct node *next;
    int value;
};

struct stack
{
    struct node *head;
};

struct stack *create_stack()
//@ requires true;
//@ ensures stack(result, nil);
{
    struct stack *stack = malloc(sizeof(struct stack));
    if (stack == 0)
    {
        abort();
    }
    stack->head = 0;
    //@ close nodes(0, nil);
    //@ close stack(stack, nil);
    return stack;
}

void stack_push(struct stack *stack, int value)
//@ requires stack(stack, ?vs);
//@ ensures stack(stack, cons(value, vs));
{
    //@ open stack(stack, vs);
    struct node *n = malloc(sizeof(struct node));
    if (n == 0)
    {
        abort();
    }
    n->next = stack->head;
    n->value = value;
    //@ close node(n, value, stack->head);
    stack->head = n;
    //@ close nodes(n, cons(value, vs));
    //@ close stack(stack, cons(value, vs));
}

int stack_pop(struct stack *stack)
//@ requires stack(stack, cons(?v, ?vs));
//@ ensures stack(stack, vs) &*& result == v;
{
    //@ open stack(stack, cons(v, vs));
    struct node *head = stack->head;
    //@ open nodes(head, cons(v, vs));
    //@ open node(head, v, ?next);
    int result = head->value;
    stack->head = head->next;
    free(head);
    //@ close nodes(stack->head, vs);
    //@ close stack(stack, vs);
    return result;
}

typedef bool int_predicate(int x);

struct node *nodes_filter(struct node *n, int_predicate *p)
//@ requires nodes(n, ?vs) &*& p(?dummy) == (dummy != dummy); // dummy to bind p's type
//@ ensures nodes(result, ?ws) &*& sublist(ws, vs) == true;
{
    if (n == 0)
    {
        //@ close nodes(0, nil);
        return 0;
    }
    else
    {
        //@ open nodes(n, vs);
        //@ open node(n, ?v, ?next);
        bool keep = p(n->value);
        if (keep)
        {
            struct node *next = nodes_filter(n->next, p);
            //@ assert nodes(next, ?filtered_next);
            n->next = next;
            //@ close node(n, v, next);
            //@ close nodes(n, cons(v, filtered_next));
            return n;
        }
        else
        {
            struct node *next = n->next;
            free(n);
            struct node *result = nodes_filter(next, p);
            return result;
        }
    }
}

void stack_filter(struct stack *stack, int_predicate *p)
//@ requires stack(stack, ?vs) &*& p(?dummy) == (dummy != dummy);
//@ ensures stack(stack, ?ws) &*& sublist(ws, vs) == true;
{
    //@ open stack(stack, vs);
    struct node *head = nodes_filter(stack->head, p);
    //@ assert nodes(head, ?ws);
    stack->head = head;
    //@ close stack(stack, ws);
}

void nodes_dispose(struct node *n)
//@ requires nodes(n, ?vs);
//@ ensures true;
{
    if (n != 0)
    {
        //@ open nodes(n, vs);
        //@ open node(n, ?v, ?next);
        nodes_dispose(n->next);
        free(n);
    }
}

void stack_dispose(struct stack *stack)
//@ requires stack(stack, ?vs);
//@ ensures true;
{
    //@ open stack(stack, vs);
    nodes_dispose(stack->head);
    free(stack);
}

bool neq_20(int x) 
//@ requires true;
//@ ensures true;
{
    return x != 20;
}

int main()
//@ requires true;
//@ ensures true;
{
    struct stack *s = create_stack();
    stack_push(s, 10);
    stack_push(s, 20);
    stack_push(s, 30);
    stack_filter(s, neq_20);
    stack_dispose(s);
    return 0;
}

File: output_log/claude-opus/sample_2/rust/claude-opus/rust/purely_unsafe/solutions/filter.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

struct Node {
    next: *mut Node,
    value: i32,
}

struct Stack {
    head: *mut Node,
}

type I32Predicate = unsafe fn(i32) -> bool;

predicate nodes(n: *mut Node;) =
    n == std::ptr::null_mut() ?
        true
    :
        (*n).next |-> ?next
        &*& (*n).value |-> ?v
        &*& malloc_block_Node(n)
        &*& nodes(next)
    ;

// ownership invariant for stack: owns head pointer and malloc block for Stack struct and linked list nodes reachable from head
predicate stack(s: *mut Stack;) =
    (*s).head |-> ?h
    &*& malloc_block_Stack(s)
    &*& nodes(h)
    ;

// abstract predicate describing ownership and permission of predicate pointer p, relates to function pointer p with shared readonly permission
predicate_ctor predicate_req(p: I32Predicate)(v: int) = 
    [_]is_I32Predicate(p) &*& true; // pure fact combined with readonly permission on p

predicate_ctor predicate_ens(p: I32Predicate)(v: int; r: bool) = 
    true; // no heap ownership so pure

#[requires(nodes(n) &*& [_]is_I32Predicate(p))]
#[ensures(nodes(result))]
unsafe fn filter_nodes(n: *mut Node, p: I32Predicate) -> *mut Node {
    if n.is_null() {
        std::ptr::null_mut()
    } else {
        let keep = p((*n).value);
        let next;
        if keep {
            next = filter_nodes((*n).next, p);

            (*n).next = next;

            n
        } else {
            next = (*n).next;
            dealloc(n as *mut u8, Layout::new::<Node>());
            let result = filter_nodes(next, p);
            result
        }
    }
}

#[requires(nodes(n))]
#[ensures(true)]
unsafe fn dispose_nodes(n: *mut Node) {
    if !n.is_null() {
        dispose_nodes((*n).next);
        dealloc(n as *mut u8, Layout::new::<Node>());
    }
}

impl Stack {
    #[ensures(stack(result))]
    unsafe fn create() -> *mut Stack {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();

        stack
    }

    #[requires(stack(stack))]
    #[ensures(stack(stack))]
    unsafe fn push(stack: *mut Stack, value: i32) {
        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }
        (*n).next = (*stack).head;
        (*n).value = value;
        (*stack).head = n;
    }

    #[requires(stack(stack))]
    #[ensures(stack(stack))]
    unsafe fn pop(stack: *mut Stack) -> i32 {
        let head = (*stack).head;

        let result = (*head).value;
        (*stack).head = (*head).next;
        dealloc(head as *mut u8, Layout::new::<Node>());

        result
    }

    #[requires(stack(stack) &*& [_]is_I32Predicate(p))]
    #[ensures(stack(stack))]
    unsafe fn filter(stack: *mut Stack, p: I32Predicate) {
        let head = filter_nodes((*stack).head, p);

        (*stack).head = head;
    }

    #[requires(stack(stack))]
    #[ensures(true)]
    unsafe fn dispose(stack: *mut Stack) {
        dispose_nodes((*stack).head);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }
}

#[ensures(true)]
unsafe fn neq_20(x: i32) -> bool {
    x != 20
}

fn main() {
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);

        Stack::filter(s, neq_20);
        Stack::dispose(s);
    }
}